# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
#Import your libraries
import numpy as np 
import pandas as pd 
from sqlalchemy import create_engine

In [2]:
pd.set_option('display.max_columns', 100)


# Challenge 1 - Import and Describe the Dataset

#### In this challenge we will use the `austin_weather` data. 

#### First, import it into a data frame called `austin`. You can find the data in Ironhack's database (db: weather, table: austin_weather). 

In [3]:
# Your code here
driver   = 'mysql+pymysql:'
user     = 'data-students'
password = 'iR0nH@cK-D4T4B4S3'
ip       = '34.65.10.136'
database = 'weather'
connection_string = f'{driver}//{user}:{password}@{ip}/{database}'
print(connection_string)
engine = create_engine(connection_string)
print(engine)
austin = pd.read_sql('SELECT * FROM austin_weather;', engine)
austin.shape

mysql+pymysql://data-students:iR0nH@cK-D4T4B4S3@34.65.10.136/weather
Engine(mysql+pymysql://data-students:***@34.65.10.136/weather)


(1319, 21)

#### Next, describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [7]:
# Your code here
austin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
Date                          1319 non-null object
TempHighF                     1319 non-null int64
TempAvgF                      1319 non-null int64
TempLowF                      1319 non-null int64
DewPointHighF                 1319 non-null object
DewPointAvgF                  1319 non-null object
DewPointLowF                  1319 non-null object
HumidityHighPercent           1319 non-null object
HumidityAvgPercent            1319 non-null object
HumidityLowPercent            1319 non-null object
SeaLevelPressureHighInches    1319 non-null object
SeaLevelPressureAvgInches     1319 non-null object
SeaLevelPressureLowInches     1319 non-null object
VisibilityHighMiles           1319 non-null object
VisibilityAvgMiles            1319 non-null object
VisibilityLowMiles            1319 non-null object
WindHighMPH                   1319 non-null object
WindAvgMPH                   

In [8]:
# Your code here
austin.describe()

TempHighF     TempAvgF     TempLowF
count  1319.000000  1319.000000  1319.000000
mean     80.862775    70.642911    59.902957
std      14.766523    14.045904    14.190648
min      32.000000    29.000000    19.000000
25%      72.000000    62.000000    49.000000
50%      83.000000    73.000000    63.000000
75%      92.000000    83.000000    73.000000
max     107.000000    93.000000    81.000000

In [18]:
# Your code here
austin.head()

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  \
0           43                  93                 75                 57   
1           28                  93                 68                 43   
2           23                  76                 52                 27   
3           21                  89                 56                 22   
4           36                  86                 71                 56   

  SeaLevelPressureHighInches SeaLevelPressureAvgInches  \
0                      29.86                     29.68   
1                      30.41                     30.13   
2                      30.56                     30.49   
3                      30.56                     30.45   
4                      30.41                     30.33   

  SeaLevelPressureLowInches VisibilityHighMiles VisibilityAvgMiles  \
0                     29.59                  10                  7   
1                     29.87                  10                 10   
2                     30.41                  10                 10   
3                      30.3                  10                 10   
4                     30.27                  10                 10   

  VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  2          20          4          31   
1                  5          16          6          25   
2                 10           8          3          12   
3                  7          12          4          20   
4                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [43]:
# Your code here
austin.DewPointAvgF.unique()

array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)

In [ ]:
# Your observation here

# there is a dash, this is why pandas treat this column as object instead of int64

In [32]:
# Your observation here
dash = austin[austin['DewPointHighF'] == '-']
dash
float_cols = ['SeaLevelPressureHighInches', 'SeaLevelPressureAvgInches', 'SeaLevelPressureLowInches', 'PrecipitationSumInches']
int_cols = austin.columns.difference(float_cols + ['Events'] + ['Date'])
int_cols

Index(['DewPointAvgF', 'DewPointHighF', 'DewPointLowF', 'HumidityAvgPercent',
       'HumidityHighPercent', 'HumidityLowPercent', 'TempAvgF', 'TempHighF',
       'TempLowF', 'VisibilityAvgMiles', 'VisibilityHighMiles',
       'VisibilityLowMiles', 'WindAvgMPH', 'WindGustMPH', 'WindHighMPH'],
      dtype='object')

In [34]:
test = austin[int_cols].apply(pd.to_numeric, errors='coerce')
test.dtypes

DewPointAvgF           float64
DewPointHighF          float64
DewPointLowF           float64
HumidityAvgPercent     float64
HumidityHighPercent    float64
HumidityLowPercent     float64
TempAvgF                 int64
TempHighF                int64
TempLowF                 int64
VisibilityAvgMiles     float64
VisibilityHighMiles    float64
VisibilityLowMiles     float64
WindAvgMPH             float64
WindGustMPH            float64
WindHighMPH            float64
dtype: object

In [41]:
test1 = austin['DewPointAvgF'].str.contains('\d', regex=True)
test1.value_counts()

True     1312
False       7
Name: DewPointAvgF, dtype: int64

In [37]:
austin[int_cols] = austin[int_cols].apply(pd.to_numeric, errors='coerce', inplace=True)


Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  \
0           43                  93                 75                 57   
1           28                  93                 68                 43   
2           23                  76                 52                 27   
3           21                  89                 56                 22   
4           36                  86                 71                 56   

  SeaLevelPressureHighInches SeaLevelPressureAvgInches  \
0                      29.86                     29.68   
1                      30.41                     30.13   
2                      30.56                     30.49   
3                      30.56                     30.45   
4                      30.41                     30.33   

  SeaLevelPressureLowInches VisibilityHighMiles VisibilityAvgMiles  \
0                     29.59                  10                  7   
1                     29.87                  10                 10   
2                     30.41                  10                 10   
3                      30.3                  10                 10   
4                     30.27                  10                 10   

  VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  2          20          4          31   
1                  5          16          6          25   
2                 10           8          3          12   
3                  7          12          4          20   
4                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T

In [ ]:
austin.filter(regex='-')

The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [4]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [5]:
# Your code here
austin[wrong_type_columns] = austin[wrong_type_columns].apply(pd.to_numeric, errors='coerce')
austin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
Date                          1319 non-null object
TempHighF                     1319 non-null int64
TempAvgF                      1319 non-null int64
TempLowF                      1319 non-null int64
DewPointHighF                 1312 non-null float64
DewPointAvgF                  1312 non-null float64
DewPointLowF                  1312 non-null float64
HumidityHighPercent           1317 non-null float64
HumidityAvgPercent            1317 non-null float64
HumidityLowPercent            1317 non-null float64
SeaLevelPressureHighInches    1316 non-null float64
SeaLevelPressureAvgInches     1316 non-null float64
SeaLevelPressureLowInches     1316 non-null float64
VisibilityHighMiles           1307 non-null float64
VisibilityAvgMiles            1307 non-null float64
VisibilityLowMiles            1307 non-null float64
WindHighMPH                   1317 non-null float64
WindAvgMPH      

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [61]:
# Your code here
austin.dtypes

Date                           object
TempHighF                       int64
TempAvgF                        int64
TempLowF                        int64
DewPointHighF                 float64
DewPointAvgF                  float64
DewPointLowF                  float64
HumidityHighPercent           float64
HumidityAvgPercent            float64
HumidityLowPercent            float64
SeaLevelPressureHighInches    float64
SeaLevelPressureAvgInches     float64
SeaLevelPressureLowInches     float64
VisibilityHighMiles           float64
VisibilityAvgMiles            float64
VisibilityLowMiles            float64
WindHighMPH                   float64
WindAvgMPH                    float64
WindGustMPH                   float64
PrecipitationSumInches        float64
Events                         object
dtype: object

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [6]:
# Your code here
missing_values = austin[austin.isnull().any(axis=1)]
missing_values.head()

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
4   2013-12-25         58        50        41           44.0          40.0   
6   2013-12-27         60        53        45           41.0          39.0   
7   2013-12-28         62        51        40           43.0          39.0   
42  2014-02-01         76        66        55           62.0          59.0   
51  2014-02-10         60        48        35           49.0          36.0   

    DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
4           36.0                 86.0                71.0                56.0   
6           37.0                 83.0                65.0                47.0   
7           33.0                 92.0                64.0                36.0   
42          41.0                 81.0                71.0                60.0   
51          30.0                 82.0                74.0                66.0   

    SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
4                        30.41                      30.33   
6                        30.46                      30.39   
7                        30.33                      30.17   
42                       29.91                      29.81   
51                       30.23                      30.15   

    SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
4                       30.27                 10.0                10.0   
6                       30.34                 10.0                 9.0   
7                       30.04                 10.0                10.0   
42                      29.75                 10.0                10.0   
51                      30.02                 10.0                 8.0   

    VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  \
4                  7.0         10.0         2.0         16.0   
6                  7.0          7.0         1.0         11.0   
7                  7.0         10.0         2.0         14.0   
42                 9.0         14.0         6.0         26.0   
51                 4.0         15.0         9.0         23.0   

    PrecipitationSumInches Events  
4                      NaN         
6                      NaN         
7                      NaN         
42                     NaN   Rain  
51                     NaN   Rain

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [74]:
# Your code here
austin.shape[0]

1319

In [75]:
missing_values.shape[0]

136

#### Calculate the ratio of missing rows to total rows

In [8]:
# Your code here
(missing_values.shape[0]/austin.shape[0])*100

10.310841546626232

As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [9]:
# Your code here
for elem in list(austin.columns):
    print(elem)
    print(austin[elem].isna().sum())
    print("······························")

Date
0
······························
TempHighF
0
······························
TempAvgF
0
······························
TempLowF
0
······························
DewPointHighF
7
······························
DewPointAvgF
7
······························
DewPointLowF
7
······························
HumidityHighPercent
2
······························
HumidityAvgPercent
2
······························
HumidityLowPercent
2
······························
SeaLevelPressureHighInches
3
······························
SeaLevelPressureAvgInches
3
······························
SeaLevelPressureLowInches
3
······························
VisibilityHighMiles
12
······························
VisibilityAvgMiles
12
······························
VisibilityLowMiles
12
······························
WindHighMPH
2
······························
WindAvgMPH
2
······························
WindGustMPH
4
······························
PrecipitationSumInches
124
······························
Events
0


#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [10]:
# Your code here
(missing_values['PrecipitationSumInches'].isna().sum()/austin['PrecipitationSumInches'].count())*100

10.376569037656903

Over 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. It doesn't make sense to *guess* its missing values because the estimation will be too 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [7]:
# Your code here 
austin.drop(['PrecipitationSumInches'],inplace=True, axis=1)

# Print `austin` to confirm the column is indeed removed

austin.dtypes

Date                           object
TempHighF                       int64
TempAvgF                        int64
TempLowF                        int64
DewPointHighF                 float64
DewPointAvgF                  float64
DewPointLowF                  float64
HumidityHighPercent           float64
HumidityAvgPercent            float64
HumidityLowPercent            float64
SeaLevelPressureHighInches    float64
SeaLevelPressureAvgInches     float64
SeaLevelPressureLowInches     float64
VisibilityHighMiles           float64
VisibilityAvgMiles            float64
VisibilityLowMiles            float64
WindHighMPH                   float64
WindAvgMPH                    float64
WindGustMPH                   float64
Events                         object
dtype: object

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [8]:
# Your code here
austin_fixed=austin.interpolate(inplace=False)

#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [8]:
# Your code here
for elem in list(austin_fixed.columns):
    print(elem)
    print(austin_fixed[elem].isna().sum())
    print("······························")

Date
0
······························
TempHighF
0
······························
TempAvgF
0
······························
TempLowF
0
······························
DewPointHighF
0
······························
DewPointAvgF
0
······························
DewPointLowF
0
······························
HumidityHighPercent
0
······························
HumidityAvgPercent
0
······························
HumidityLowPercent
0
······························
SeaLevelPressureHighInches
0
······························
SeaLevelPressureAvgInches
0
······························
SeaLevelPressureLowInches
0
······························
VisibilityHighMiles
0
······························
VisibilityAvgMiles
0
······························
VisibilityLowMiles
0
······························
WindHighMPH
0
······························
WindAvgMPH
0
······························
WindGustMPH
0
······························
Events
0
······························


In [15]:
for elem in list(austin.columns):
    print(elem)
    print(austin[elem].isna().sum())
    print("······························")

Date
0
······························
TempHighF
0
······························
TempAvgF
0
······························
TempLowF
0
······························
DewPointHighF
7
······························
DewPointAvgF
7
······························
DewPointLowF
7
······························
HumidityHighPercent
2
······························
HumidityAvgPercent
2
······························
HumidityLowPercent
2
······························
SeaLevelPressureHighInches
3
······························
SeaLevelPressureAvgInches
3
······························
SeaLevelPressureLowInches
3
······························
VisibilityHighMiles
12
······························
VisibilityAvgMiles
12
······························
VisibilityLowMiles
12
······························
WindHighMPH
2
······························
WindAvgMPH
2
······························
WindGustMPH
4
······························
Events
0
······························


# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [9]:
# Your code here:
austin_fixed.Events.value_counts()

                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: Events, dtype: int64

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [ ]:
# Your answer:
nothing (903) followed by Rain(192)

#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [10]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

# Your code here
for elem in event_list:
    austin_fixed[elem]=0

# Print your new dataframe to check whether new columns have been created:

austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH     

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [11]:
# Your code here
for elem in event_list:
    austin_fixed[elem]=austin_fixed.Events.str.contains(elem, regex=False).astype(int)

#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [23]:
# Your code here
austin_fixed.sample(20)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
1276  2017-06-19         92        84        76           74.0          72.0   
1168  2017-03-03         71        56        40           38.0          29.0   
459   2015-03-25         84        73        62           63.0          61.0   
904   2016-06-12         94        86        77           77.0          76.0   
66    2014-02-25         71        62        52           65.0          60.0   
966   2016-08-13         96        87        77           76.0          73.0   
922   2016-06-30         94        85        76           74.0          71.0   
456   2015-03-22         78        68        57           59.0          57.0   
1186  2017-03-21         85        74        63           63.0          61.0   
776   2016-02-05         62        48        33           39.0          28.0   
577   2015-07-21         97        86        75           73.0          71.0   
939   2016-07-17         99        88        76           74.0          72.0   
21    2014-01-11         75        62        48           54.0          37.0   
1302  2017-07-15        103        87        71           73.0          70.0   
519   2015-05-24         82        73        64           70.0          68.0   
550   2015-06-24         92        83        74           74.0          73.0   
362   2014-12-18         60        55        49           56.0          51.0   
538   2015-06-12         93        84        75           75.0          73.0   
1123  2017-01-17         59        54        48           55.0          48.0   
780   2016-02-09         74        57        39           33.0          27.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
1276          68.0                 94.0                72.0   
1168          24.0                 70.0                45.0   
459           59.0                 93.0                72.0   
904           75.0                 91.0                74.0   
66            47.0                 93.0                85.0   
966           70.0                 94.0                73.0   
922           67.0                 88.0                65.0   
456           55.0                100.0                75.0   
1186          59.0                 93.0                70.0   
776           23.0                 70.0                47.0   
577           66.0                 94.0                66.0   
939           66.0                 88.0                63.0   
21            20.0                 93.0                53.0   
1302          63.0                100.0                65.0   
519           63.0                 93.0                81.0   
550           70.0                100.0                78.0   
362           48.0                100.0                92.0   
538           71.0                 88.0                70.0   
1123          46.0                100.0                86.0   
780           20.0                 59.0                40.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
1276                49.0                       30.03   
1168                19.0                       30.57   
459                 51.0                       29.98   
904                 56.0                       30.09   
66                  77.0                       30.06   
966                 52.0                       29.88   
922                 41.0                       29.99   
456                 50.0                       30.08   
1186                47.0                       30.14   
776                 24.0                       30.48   
577                 38.0                       29.95   
939                 38.0                       30.08   
21                  13.0                       30.16   
1302                29.0                       30.07   
519                 69.0                       29.95   
550                 56.0                       30.19   
362                 84.0                       30.14   

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [12]:
# Your code here
austin_fixed.drop(['Events'],inplace=True,axis=1)

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [13]:
from datetime import datetime as dt

In [14]:
# Your code here
austin_fixed.Date = austin_fixed['Date'].apply(lambda x: dt.strptime(x, '%Y-%m-%d').toordinal())

In [44]:
# Your code here
for x in list(austin_fixed.Date):
    print(dt.strptime(x, '%Y-%m-%d').toordinal())

735223
735224
735225
735226
735227
735228
735229
735230
735231
735232
735233
735234
735235
735236
735237
735238
735239
735240
735241
735242
735243
735244
735245
735246
735247
735248
735249
735250
735251
735252
735253
735254
735255
735256
735257
735258
735259
735260
735261
735262
735263
735264
735265
735266
735267
735268
735269
735270
735271
735272
735273
735274
735275
735276
735277
735278
735279
735280
735281
735282
735283
735284
735285
735286
735287
735288
735289
735290
735291
735292
735293
735294
735295
735296
735297
735298
735299
735300
735301
735302
735303
735304
735305
735306
735307
735308
735309
735310
735311
735312
735313
735314
735315
735316
735317
735318
735319
735320
735321
735322
735323
735324
735325
735326
735327
735328
735329
735330
735331
735332
735333
735334
735335
735336
735337
735338
735339
735340
735341
735342
735343
735344
735345
735346
735347
735348
735349
735350
735351
735352
735353
735354
735355
735356
735357
735358
735359
735360
735361
735362
735363
735364
735365

#### Print `austin_fixed` to check your `Date` column.

In [19]:
austin_fixed.head(5)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  735223         74        60        45           67.0          49.0   
1  735224         56        48        39           43.0          36.0   
2  735225         58        45        32           31.0          27.0   
3  735226         61        46        31           36.0          28.0   
4  735227         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
0                      29.59                 10.0                 7.0   
1                      29.87                 10.0                10.0   
2                      30.41                 10.0                10.0   
3                      30.30                 10.0                10.0   
4                      30.27                 10.0                10.0   

   VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  Rain  \
0                 2.0         20.0         4.0         31.0     0    0     1   
1                 5.0         16.0         6.0         25.0     0    0     0   
2                10.0          8.0         3.0         12.0     0    0     0   
3                 7.0         12.0         4.0         20.0     0    0     0   
4                 7.0         10.0         2.0         16.0     0    0     0   

   Thunderstorm  
0             1  
1             0  
2             0  
3             0  
4             0

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [22]:
# Your code here:
x = austin_fixed[['TempHighF','TempLowF','DewPointHighF','DewPointAvgF','DewPointLowF','HumidityHighPercent','HumidityAvgPercent','HumidityLowPercent','SeaLevelPressureHighInches','SeaLevelPressureAvgInches','SeaLevelPressureLowInches','VisibilityHighMiles','VisibilityAvgMiles','VisibilityLowMiles','WindHighMPH','WindAvgMPH','WindGustMPH','Snow','Fog','Rain','Thunderstorm']]
y = austin_fixed['TempAvgF']

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [24]:
#Your code here:
from sklearn.model_selection import train_test_split

Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [25]:
#Your code here:
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2)

#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [ ]:
# Your code here:


Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [ ]:
# Your code here:


Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [ ]:
# Your code here:
